### PR TITLE
Spring Cloud property view

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/UIHelperImpl.java [hunk 1]
```diff
@@ -97,6 +101,9 @@
     public static final Key<String> RESOURCE_ID = new Key<>("resourceId");
     public static final Key<String> WEBAPP_ID = new Key<>("webAppId");
     public static final Key<String> APP_ID = new Key<>("appId");
+    public static final Key<AppResourceInner> SPRING_CLOUD_APP = new Key<>("springCloudApp");
+    public static final Key<DeploymentResourceInner> SPRING_CLOUD_DEPLOYMENT = new Key<>("springCloudAppDeployment");
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/UIHelperImpl.java:105]: is `SPRING_CLOUD_DEPLOYMENT` used somewhere?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringAppViewModel.java [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+    public void setUsedStorageInGB(Integer usedStorageInGB) {
+        this.usedStorageInGB = usedStorageInGB;
+    }
+
+    public boolean isCanStart() {
+        return canStart;
+    }
+
+    public void setCanStart(boolean canStart) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringAppViewModel.java:220]: do we need to add `canStart` `canStop` methods in Model? As these valued could get from status

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+        instancesTableModel.addColumn("Status");
+        instancesTableModel.addColumn("Discover Status");
+        instanceTable.setModel(instancesTableModel);
+        instanceTable.setRowSelectionAllowed(true);
+        instanceTable.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
+        instanceTable.getEmptyText().setText("Loading instances status");
+        instanceTable.setPreferredSize(new Dimension(-1, 200));
+        this.saveButton.addActionListener(e -> {
+            wrapperOperations(TelemetryConstants.SAVE_SPRING_CLOUD_APP, project, (changes) -> {
+                if (changes.isEmpty()) {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java:144]: when will changes empty but save button enabled?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+            this.model.setStatus(status.toString());
+            this.setData(model);
+        } catch (AzureExecutionException e) {
+            ApplicationManager.getApplication().invokeLater(() -> {
+                PluginUtil.showErrorNotificationProject(project, "Cannot binding data to Spring Cloud property view.", e.getMessage());
+            });
+        }
+    }
+
+    private void setData(SpringAppViewModel model) throws AzureExecutionException {
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java:621]: setData -> refreshUI? As setData seems more likely an Model operation

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+                publicUrlHyperLink.setHyperlinkTarget(publicUrl);
+                Font font = publicUrlHyperLink.getFont();
+                publicUrlHyperLink.setFont(new Font(font.getName(), Font.PLAIN, font.getSize()));
+            } else {
+                publicUrlHyperLink.setText(publicUrl);
+                Font font = publicUrlHyperLink.getFont();
+                publicUrlHyperLink.setFont(new Font(font.getName(), Font.ITALIC, font.getSize()));
+            }
+        } else {
+            publicUrlHyperLink.setText(NOT_AVAILABLE);
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java:514]: add setHyperlinkTarget to empty

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/UIHelperImpl.java (hunk 1 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/UIHelperImpl.java]: No need for this const

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringAppViewModel.java (hunk 2 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringAppViewModel.java:220]: will consider to refactor it later

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java (hunk 3 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java:147]: There are some chances that dirty check is not performed on time.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java (hunk 4 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java]: setData is changed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java (hunk 5 above)
Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/helpers/springcloud/SpringCloudAppPropertyView.java:522]: done

**Pull request merged**